### ISSUE TITLE
RegEx Revamp Ideas

### ISSUE THREAD

**Issue by leezh** [godotengine/godot#6853]
I have been mulling over converting the regex library into a module in IRC, and I thought it would be a good opportunity to add to the library, one of which is going to break backwards compatibility. What do you think of them? Any ideas you'd like to add / modify to this?
### Results in a separate object

Previously, the results would be stored as a mutable variable in the expression object. I don't think it's a good principle to have a constant function `ex.find()` to affect the state of `ex`. I originally did this because that was how it was done and I was worried about backwards compatibility. What I propose is to instead return a separate object called RegExMatch, similar to [Python](https://docs.python.org/3/library/re.html):

Current:

```
var ex = RegEx.new()
ex.compile("Name: (\\w+)")
if ex.find("ID: 003\nName: John Doe"):
    print(ex.get_capture(1))
    print(ex.get_capture_start(1))
```

New (Proposed):

```
var ex = RegEx.new()
ex.compile("Name: (\\w+)")
var match = ex.match("ID: 003\nName: John Doe")
if match != null:
    print(match.get_group(1))
    print(match.get_pos(1))
```
### RegEx as a resource file

One annoyance I've had making use of the RegEx class is writing escapes because I have to escape the string before I can escape the expression, e.g. "\w" -> \w or worse "\\" -> \. Furthermore, the lack of access to constructors in the script means if you want to use them as a class member, it has to be compiled separately:

```
var re_string = RegEx.new()
var re_comments = RegEx.new()
# Some code
func _init():
    re_string.compile("\"([^\"]|\\\\.)*\"")
    re_comments.compile("#.*")
# Some more code
func _input(delta):
    # Actually use re_string and re_comments
    pass
```

One way to fix all this is to make it a resource class instead, where you can edit it in the properties window, save it as a file, and load it separately, like:

```
var re_string = load("res://regex/string.tres")
func _input(event):
    # With resource caching, it should only be loaded once
    var re_string2 = load("res://regex/string.tres")
    # The old method should still work
    var re_comments = RegEx.new()
    re_comments.compile("#.*")
```
### Other ideas

By integrating more closely with Godot, it should now be possible to add:
- Substitution
- Named groups

**Comment by reduz** [comment]
Ive been told the Regex module uses dynamic_cast
since we disable RTTI in a few platforms due to size constraint, I guess we
should make it work without it

**Comment by bojidar-bg** [comment]
I think it is fixed in a recent PR, #6958
